src/register/macros.rs

```diff
@@ -1,255 +1,198 @@
-macro_rules! read_csr {
-    ($csr_number:literal) => {
-        /// Reads the CSR
-        #[inline]
-        unsafe fn _read() -> usize {
-            match () {
-                #[cfg(riscv)]
-                () => {
-                    let r: usize;
-                    core::arch::asm!(concat!("csrrs {0}, ", stringify!($csr_number), ", x0"), out(reg) r);
-                    r
-                }
-
-                #[cfg(not(riscv))]
-                () => unimplemented!(),
-            }
-        }
-    };
-}
-
-macro_rules! read_csr_rv32 {
-    ($csr_number:literal) => {
-        /// Reads the CSR
-        #[inline]
-        unsafe fn _read() -> usize {
-            match () {
-                #[cfg(riscv32)]
-                () => {
-                    let r: usize;
-                    core::arch::asm!(concat!("csrrs {0}, ", stringify!($csr_number), ", x0"), out(reg) r);
-                    r
-                }
-
-                #[cfg(not(riscv32))]
-                () => unimplemented!(),
-            }
-        }
-    };
-}
+// Copyright (c) 2022 by Rivos Inc.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
 
-macro_rules! read_csr_as {
-    ($register:ident, $csr_number:literal) => {
-        read_csr!($csr_number);
+//! Various macros to help with creating CSR modules
 
-        /// Reads the CSR
-        #[inline]
-        pub fn read() -> $register {
-            $register {
-                bits: unsafe { _read() },
-            }
+/// This macro generates the constants and types for a read-write CSR.
+macro_rules! rw_csr {
+    // Use this if the CSR name can be matched exactly with the bitfield
+    // generated using `register_bitfields!` Most CSRs can use this version.
+    ($name:ident, $size:ty) => {
+        paste::paste! {
+            rw_csr!($name, [< $name:lower >], $size);
         }
     };
-}
-
-macro_rules! read_csr_as_usize {
-    ($csr_number:literal) => {
-        read_csr!($csr_number);
-
-        /// Reads the CSR
-        #[inline]
-        pub fn read() -> usize {
-            unsafe { _read() }
-        }
+    // This version will need to be used if the CSR layout is not unique. An
+    // example would be the PMP CSRs, there are multiple with the exact same
+    // layout. Using this allows the same bitfield to be used multiple times.
+    // Ex:
+    //   rw_csr!(pmpcfg0, pmpcfg, usize);
```

Comment:
what type is `pmpcfg` in this example? is it generated by `rw_csr`?

Reply:
Furquan is right, it would come from `register_bitfields`. I'll make that clearer in the documentation.